afpacket/afpacket.go

```diff
@@ -248,16 +250,17 @@ func (h *TPacket) releaseCurrentPacket() error {
 func (h *TPacket) ZeroCopyReadPacketData() (data []byte, ci gopacket.CaptureInfo, err error) {
 	h.mu.Lock()
 retry:
-	if h.current == nil || !h.current.next() {
+	if h.current == nil || !h.headerNextNeeded || !h.current.next() {
 		if h.shouldReleasePacket {
 			h.releaseCurrentPacket()
 		}
 		h.current = h.getTPacketHeader()
```

Comment:
I think you've definitely found some code that needs some work, but I'm not sure if this solution actually addresses the main concern I have with it, after looking closely.

From what I can see, the problem isn't really with the .next() call (althought that is an issue), it's actually with the getTPacketHeader() call, which advances h.current to the next block.  Should poll error out, there's a chance that we'll miss packets, since h.getTPacketHeader() will be called twice, advancing _past_ a block that could have packets in it.

I'm going to go back and take a look at this code myself, and maybe together we can rejigger things to handle poll errors better.  Thanks for bringing this up!